### PR TITLE
Fix ClassCastException in toBeamValue with List<{Primitive}> types

### LINKED ISSUE #26870: [Bug]: toBeamValue in BigQueryUtils throws ClassCastException when converting ARRAY<STRING>
### What happened?

On 2.48.0-SNAPSHOT, reading from a BigQuery table with a column of type `ARRAY<STRING NOT NULL> NOT NULL` causes the following exception:
```
org.apache.beam.sdk.Pipeline$PipelineExecutionException: java.lang.ClassCastException: class java.lang.String cannot be cast to class java.util.Map (java.lang.String and java.util.Map are in module java.base of loader 'bootstrap')

	at org.apache.beam.runners.direct.DirectRunner$DirectPipelineResult.waitUntilFinish(DirectRunner.java:374)
	at org.apache.beam.runners.direct.DirectRunner$DirectPipelineResult.waitUntilFinish(DirectRunner.java:342)
	at com.google.cloud.syndeo.perf.SyndeoTestingPairsIT.run(SyndeoTestingPairsIT.java:216)
	at com.google.cloud.syndeo.perf.SyndeoTestingPairsIT.testSyndeoTemplateDataflow(SyndeoTestingPairsIT.java:175)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.apache.beam.sdk.testing.TestPipeline$1.evaluate(TestPipeline.java:323)
	at org.apache.beam.sdk.testing.TestPipeline$1.evaluate(TestPipeline.java:323)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:27)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:69)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater$1.execute(IdeaTestRunner.java:38)
	at com.intellij.rt.execution.junit.TestsRepeater.repeat(TestsRepeater.java:11)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:35)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:235)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:54)
Caused by: java.lang.ClassCastException: class java.lang.String cannot be cast to class java.util.Map (java.lang.String and java.util.Map are in module java.base of loader 'bootstrap')
	at org.apache.beam.sdk.io.gcp.bigquery.BigQueryUtils.lambda$toBeamValue$12(BigQueryUtils.java:697)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1655)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
	at org.apache.beam.sdk.io.gcp.bigquery.BigQueryUtils.toBeamValue(BigQueryUtils.java:699)
	at org.apache.beam.sdk.io.gcp.bigquery.BigQueryUtils.toBeamRowFieldValue(BigQueryUtils.java:643)
	at org.apache.beam.sdk.io.gcp.bigquery.BigQueryUtils.lambda$toBeamRow$6(BigQueryUtils.java:629)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1655)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
	at org.apache.beam.sdk.io.gcp.bigquery.BigQueryUtils.toBeamRow(BigQueryUtils.java:630)
	at org.apache.beam.sdk.io.gcp.bigquery.providers.BigQueryStorageWriteApiSchemaTransformProvider$BigQueryStorageWriteApiPCollectionRowTupleTransform.lambda$expand$96381ff0$1(BigQueryStorageWriteApiSchemaTransformProvider.java:348)
	at org.apache.beam.sdk.transforms.MapElements$2.processElement(MapElements.java:151)
```

Looking at `BigQueryUtils.java:697`, it looks like `toBeamValue` expects a List to have inner type `Map<String, Object>`, when the passed in value is `List<String>`.

### Issue Priority

Priority: 2 (default / most bugs should be filed as P2)

### Issue Components

- [ ] Component: Python SDK
- [ ] Component: Java SDK
- [ ] Component: Go SDK
- [ ] Component: Typescript SDK
- [X] Component: IO connector
- [ ] Component: Beam examples
- [ ] Component: Beam playground
- [ ] Component: Beam katas
- [ ] Component: Website
- [ ] Component: Spark Runner
- [ ] Component: Flink Runner
- [ ] Component: Samza Runner
- [ ] Component: Twister2 Runner
- [ ] Component: Hazelcast Jet Runner
- [ ] Component: Google Cloud Dataflow Runner

### PR DESCRIPTION
**Pull request by jonathan-lemos**
Fixes #26870

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by jonathan-lemos** [comment]
R: @ahmedabu98

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryUtils.java [hunk 1]
```diff
@@ -694,7 +694,7 @@ public static Row toBeamRow(Schema rowSchema, TableSchema bqSchema, TableRow jso
     if (jsonBQValue instanceof List) {
       return ((List<Object>) jsonBQValue)
           .stream()
-              .map(v -> ((Map<String, Object>) v).get("v"))
+              .map(v -> v instanceof Map ? ((Map<String, Object>) v).get("v") : v)
```
[the remark this answers is not in the document]
Comment (jonathan-lemos) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryUtils.java]: I have covered this case.
Interestingly, it doesn't look like BigQuery supports a Map type (https://cloud.google.com/bigquery/docs/reference/standard-sql/data-types#data_type_sizes), so I don't know how to test this path.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryUtils.java [hunk 2]
```diff
@@ -694,7 +694,16 @@ public static Row toBeamRow(Schema rowSchema, TableSchema bqSchema, TableRow jso
...
+              .map(
+                  v -> {
+                    if (v instanceof Map) {
+                      Map<String, Object> m = (Map<String, Object>) v;
+                      if (m.size() == 1 && m.containsKey("v")) {
+                        return m.get("v");
+                      }
+                    }
+                    return v;
+                  })
```
Comment (ahmedabu98) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryUtils.java:706]: I was thinking something like this, just to check if the user is intentionally using a list of maps.

```suggestion
              .map(v -> (v instanceof Map && !fieldType.getCollectionElementType().getTypeName().equals(TypeName.MAP)) ? ((Map<String, Object>) v).get("v") : v)
```

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryUtils.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (ahmedabu98) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryUtils.java]: The Map refers to the Beam Schema type, I think the equivalent for BigQuery schema might be a struct

**Review by ahmedabu98 — CHANGES_REQUESTED** [review]
added a comment

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryUtils.java (hunk 2 above)
Comment (jonathan-lemos) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryUtils.java]: That makes more sense. I've applied the change and formatted it.

Trying to make a test case for this scenario, it seems that Beam's BigQuery IO will flatten any list of structs e.g. querying
```
[
{make: honda, metadata: [{color: red, model: civic}, {color: blue, model: accord}]}
{make: toyota, metadata: [{color: green, model: corolla}]}
]
```
will result in the following:
```
[
{make: honda, metadata_color: red, metadata_model: civic}
{make: honda, metadata_color: blue, metadata_model: accord}
{make: toyota, metadata_color: green, metadata_model: corolla}
]
```

The underlying BigQuery API returns the two rows above, but Beam's [`readTableRows()`](https://beam.apache.org/releases/javadoc/current/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.html#readTableRows--) will flatten the structure and return the 3 below.

Given that behavior, it doesn't seem like it's possible for a value of `List<Struct>` to make it to this function, so I haven't added a test case for that scenario.

Comment (ahmedabu98) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryUtils.java]: Users can set this option to get results without flattening:
https://github.com/apache/beam/blob/6865eefac5633198c71fd1ec75b514e8a390d002/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java#L823

A test for List<Map<>> should be added at some point, but let's not block this PR for that

**Review by ahmedabu98 — APPROVED** [review]
LGTM,

@Abacn  can you merge?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryUtils.java [hunk 3]
```diff
@@ -694,7 +694,15 @@ public static Row toBeamRow(Schema rowSchema, TableSchema bqSchema, TableRow jso
     if (jsonBQValue instanceof List) {
       return ((List<Object>) jsonBQValue)
           .stream()
-              .map(v -> ((Map<String, Object>) v).get("v"))
+              .map(
+                  v ->
+                      (v instanceof Map
+                              && !fieldType
+                                  .getCollectionElementType()
```
Comment (Abacn) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryUtils.java:701]: are their NPE concern? fieldType.getCollectionElementType is nullable

Comment (Abacn) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryUtils.java:701]: and `!fieldType.getCollectionElementType().getTypeName().equals(TypeName.MAP)` does not depend on and can be extract outside the lambda (for visibility)

**Review by Abacn — APPROVED** [review]
LGTM. CI is rebooting. will merge once jenkins back

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryUtils.java (hunk 3 above)
Comment (jonathan-lemos) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryUtils.java]: `fieldType.getCollectionElementType()` should only return `null` if the fieldType is not a collection, but the `if (jsonBQValue instanceof List)` case should cover that scenario. However, there's nothing stopping someone from passing in a non-collection FieldType with a List value, so my latest changes show a more helpful error message in that case.

I've extracted the `!fieldType.getCollectionElementType().getTypeName().equals(TypeName.MAP)` to a variable outside of the lambda.

**Comment by Abacn** [comment]
retest this please

**Comment by Abacn** [comment]
Run RAT PreCommit

**Comment by Abacn** [comment]
Run Spotless PreCommit

**Comment by Abacn** [comment]
Run Kotlin_Examples PreCommit

**Pull request merged**